### PR TITLE
Improve validation of number of arguments passed into a function. Improve support for functions with variable length arguments

### PR DESCRIPTION
**Pull request by izeigerman**
This is a proposal to improve the validation for number of arguments passed into functions. With this change the following invalid expression fails:
```python
>>> import sqlglot
>>> sqlglot.parse("IF(a > 0, a, b, c)")
...
sqlglot.errors.ParseError: The number of provided arguments (4) is greater than the maximum number of supported arguments (3). Line 0, Col: 17.
IF(a > 0, a, b, c)
```
We still preserve support for expressions with variable length arguments:
```python
>>> sqlglot.parse("ARRAY(a, b, c)")
expressions
[(COLUMN this: a, db: , table: , fields: ), (COLUMN this: b, db: , table: , fields: ), (COLUMN this: c, db: , table: , fields: )]
[(ARRAY expressions:
  (COLUMN this: a, db: , table: , fields: ),
  (COLUMN this: b, db: , table: , fields: ),
  (COLUMN this: c, db: , table: , fields: ))]
 ```
 
Another positive impact of this change is that errors that were previously confusing became much cleaner. For example:
```python
>>> sqlglot.parse("IF(a > 0)")
...
IndexError: list index out of range
```
Note how all the useful debug information was completely lost. Here's how it looks like with the proposed change:
```python
>>> sqlglot.parse("IF(a > 0)")
...
sqlglot.errors.ParseError: Required keyword: true missing for TokenType.FUNC. Line 0, Col: 8.
IF(a > 0)
```

Additionally this change eliminates quite a bit of boilerplate associated with mapping arguments to their corresponding keys using their respective indexes. For the vast majority of cases using the new `from_arg_list` constructor as is is sufficient. Only few cases require special handling. 

Please note that the proposed change requires argument types defined in the expression to be ordered. For this purpose I introduced an additional attribute `ordered_argument_types` for derivatives of the `Func` expression only.

Please also note that I haven't added any additional tests to validate the handling of the invalid number of arguments. I'll do this once we settle on proposed changes.

### CODE REVIEW

File: sqlglot/dialects.py [hunk 1]
```diff
@@ -113,18 +113,18 @@ def _unix_to_time(self, expression):
     }
 
     functions = {
-        'APPROX_COUNT_DISTINCT': lambda args: exp.ApproxDistinct(this=args[0]),
-        'EPOCH': lambda args: exp.TimeToUnix(this=args[0]),
+        'APPROX_COUNT_DISTINCT': exp.ApproxDistinct.from_arg_list,
+        'EPOCH': exp.TimeToUnix.from_arg_list,
         'EPOCH_MS': lambda args: exp.UnixToTime(this=exp.Slash(
-            this=args[0],
+            this=list_get(args, 0),
```
Comment (izeigerman) [inline sqlglot/dialects.py:119]: @tobymao here and below I switch from accessing an argument by index directly to using the `list_get` help to avoid the generic `IndexError: list index out of range` error which leads to a loss of the parser debugging info.

Comment (izeigerman) [reply sqlglot/dialects.py:119]: Now I'm thinking it was the wrong thing to do provided the existing validation only checks whether the key was set and not whether it was None or not ([code](https://github.com/izeigerman/sqlglot/blob/main/sqlglot/expressions.py#L79))

Comment (tobymao) [reply sqlglot/dialects.py:119]: it should probably be fine?

File: sqlglot/dialects.py [hunk 2]
```diff
@@ -206,33 +206,37 @@ def ts_or_ds_to_date(self, expression):
...
+        'APPROX_COUNT_DISTINCT': exp.ApproxDistinct.from_arg_list,
+        'COLLECT_LIST': exp.ArrayAgg.from_arg_list,
+        'DATE_ADD': lambda args: exp.DateAdd.from_arg_list([exp.DateStrToDate(this=list_get(args, 0)), *args[1:]]),
+        'DATEDIFF': lambda args: exp.DateDiff.from_arg_list([
+            exp.DateStrToDate(this=list_get(args, 0)),
+            exp.DateStrToDate(this=list_get(args, 1)),
+            *args[2:]
+        ]),
+        'DATE_SUB': lambda args: exp.DateAdd.from_arg_list([
+            exp.DateStrToDate(this=list_get(args, 0)),
```
Comment (tobymao) [inline sqlglot/dialects.py:218]: is from_arg_list really more clear?

Comment (tobymao) [reply sqlglot/dialects.py:218]: what do you think about for any default version, we use from_arg_list, but if you have to change order then do it the old way?

File: sqlglot/dialects.py [hunk 3]
```diff
@@ -206,33 +206,37 @@ def ts_or_ds_to_date(self, expression):
...
+            *args[2:]
+        ]),
+        'DATE_FORMAT': exp.TimeToStr.from_arg_list,
+        'DAY': lambda args: exp.Day.from_arg_list([exp.TsOrDsToDate(this=list_get(args, 0)), *args[1:]]),
+        'FROM_UNIXTIME': lambda args: exp.UnixToStr.from_arg_list([
+            list_get(args, 0),
+            list_get(args, 1) or Hive.TIME_FORMAT,
+            *args[2:]
+        ]),
+        'GET_JSON_OBJECT': lambda args: exp.JSONPath.from_arg_list([list_get(args, 0), list_get(args, 1), *args[3:]]),  # Ignore the 3rd argument
```
Comment (tobymao) [inline sqlglot/dialects.py:229]: what do this mean? ignore the 3rd argument, why do you even have to do anything?

File: sqlglot/expressions.py [hunk 4]
```diff
@@ -484,43 +484,78 @@ class Interval(Expression):
 # Functions
 class Func(Expression):
     token_type = TokenType.FUNC
+    ordered_arg_types = [('this', True)]
+    is_var_len_args = False
+
+    def __init__(self, **args):
+        self.arg_types = dict(self.ordered_arg_types)
```
Comment (tobymao) [inline sqlglot/expressions.py:491]: it was designed specifically to have arg_types so that we don't have to worry about different dialects having different orders in functions and that with kwargs it's very explicit what each arg is, it looks like this change is making all functions use ordered_arg_types? 

i still think regular args are better and that ordered_arg_types should be optional

File: sqlglot/expressions.py [hunk 5]
```diff
@@ -484,43 +484,78 @@ class Interval(Expression):
...
+        super().__init__(**args)
+
+    @classmethod
+    def from_arg_list(cls, args):
+        args_num = len(args)
+
+        # If this function supporst variable length argument treat the last argument as such.
```
Comment (tobymao) [inline sqlglot/expressions.py:498]: supports

File: sqlglot/expressions.py [hunk 6]
```diff
@@ -484,43 +484,78 @@ class Interval(Expression):
...
+        arg_types = cls.ordered_arg_types[:-1] if cls.is_var_len_args else cls.ordered_arg_types
+
+        args_dict = {}
+        arg_idx = 0
+        for arg_key, _ in arg_types:
```
Comment (tobymao) [inline sqlglot/expressions.py:503]: enumerate(arg_types) instead of manual arg_index?

File: sqlglot/dialects.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/dialects.py:119]: Not really. The semantics has changed. Previously we failed with `IndexError`. Now we don't, since the `list_get` returns `None` if the argument is missing. However the existing validation only catches missing arguments if the key is missing (and ignores the scenario where it's set to `None` explicitly). With this change we always explicitly set the key even if the argument is missing in the list.

File: sqlglot/dialects.py (hunk 2 above)
Comment (izeigerman) [reply sqlglot/dialects.py:218]: That's a good question and I was thinking about this myself. I concluded that the `from_arg_list` is still a better option since it protects against excessive arguments out of the box. What do you think?

File: sqlglot/dialects.py (hunk 3 above)
Comment (izeigerman) [reply sqlglot/dialects.py:229]: Check out the following test case - [link](https://github.com/tobymao/sqlglot/blob/main/tests/test_dialects.py#L340-L345). The 3rd argument is indeed ignored there. The reason we pass `, *args[3:]` is to make sure there are no excessive arguments beyond the 3rd one.

File: sqlglot/expressions.py (hunk 4 above)
Comment (izeigerman) [reply sqlglot/expressions.py:491]: Makes sense. Will update this.

File: sqlglot/expressions.py (hunk 6 above)
Comment (izeigerman) [reply sqlglot/expressions.py:503]: Good point 👍 I forgot the fact that in Python whatever is defined in the `for`'s scope is available outside the block.

Comment (izeigerman) [reply sqlglot/expressions.py:503]: Provided I use `arg_idx` outside the `for` block, `enumerate` doesn't really do the right thing. Example:
```
>>> idx = 0
>>> for idx in range(0, 1):
...     pass
...
>>> idx
0
```
My expectation is that the `arg_idx` should point to the next index after the last successful iteration, not the index of the last successful iteration itself.

File: sqlglot/dialects.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/dialects.py:119]: Additionally I'm overall confused by how missing arguments are handled. For example in for the [Join](https://github.com/tobymao/sqlglot/blob/main/sqlglot/expressions.py#L234) expression we expect the `on` argument to be mandatory. However we violate this assumption when we handle `CROSS JOIN`. We work this around by actually providing the `on` key, but setting it to `None`.

File: sqlglot/expressions.py (hunk 4 above)
Comment (izeigerman) [reply sqlglot/expressions.py]: Simplified the implementation as per our conversation earlier.

File: sqlglot/expressions.py (hunk 5 above)
Comment (izeigerman) [reply sqlglot/expressions.py]: Fixed

File: sqlglot/dialects.py (hunk 3 above)
Comment (tobymao) [reply sqlglot/dialects.py:229]: ok, i think i found a cleaner solution, let me know what you think

```diff --git a/sqlglot/parser.py b/sqlglot/parser.py
index e002c92..f6af8de 100644
--- a/sqlglot/parser.py
+++ b/sqlglot/parser.py
@@ -802,7 +802,9 @@ class Parser:
             elif not callable(function):
                 this = exp.Anonymous(this=this, expressions=args)
             else:
-                this = function(args)
+                this = function(*args)
+                if len(args) > len(this.arg_types):
+                    self.raise_error('too many args')
```

two possible changes

first change, splat out args, and then when you define lambdas, you can do like
lambda x, y, *args: ... 
instead of having to do this *args[3:], really don't like that, you can also name the arguments coming in so it's more readable

also, in here, we can easily raise exception when extranous args are passed notice i can trivially check the initial arg length and compare it with the return type arg length

Comment (tobymao) [reply sqlglot/dialects.py:229]: obviously this example code doesn't take into account varargs, so it breaks on arrays

File: sqlglot/dialects.py (hunk 1 above)
Comment (tobymao) [reply sqlglot/dialects.py:119]: i think missing arguments just means that it's required to set it. i think you may be able to change this to check that the value is not none? for example

args.get(key) is not None

File: sqlglot/dialects.py (hunk 2 above)
Comment (tobymao) [reply sqlglot/dialects.py:218]: with my snippet below, i don't think it's necessary anymore for the excessive argument check

File: sqlglot/dialects.py (hunk 3 above)
Comment (izeigerman) [reply sqlglot/dialects.py:229]: I think I like the idea of moving the error handling, but not sure about the lambda x, y, *args: .... If there are not enough arguments the error will still be confusing:
```python
>>> f = lambda a, b, *args: print(a)
>>> f(*[1])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: <lambda>() missing 1 required positional argument: 'b'
```

File: sqlglot/dialects.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/dialects.py:119]: I've tried that and a bunch of tests failed :) 
```
======================================================================
ERROR: test_spark (tests.test_dialects.TestDialects)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/izeigerman/github/sqlglot/tests/test_dialects.py", line 517, in test_spark
    transpile(
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 24, in transpile
    for expression in parse(code, read, error_level=error_level)
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 13, in parse
    return dialect.parse(code, **opts)
  File "/Users/izeigerman/github/sqlglot/sqlglot/dialects.py", line 17, in parse
    return self.parser(**opts).parse(self.tokenizer().tokenize(code), code)
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 193, in parse
    expressions.append(self._parse_statement())
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 261, in _parse_statement
    cte = self._parse_cte()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 416, in _parse_cte
    return exp.CTE(
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 14, in __init__
    self.validate()
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 81, in validate
    raise ValueError(f"Required keyword: {k} missing for {self.token_type}")
ValueError: Required keyword: this missing for TokenType.WITH

======================================================================
ERROR: test_ctas (tests.test_rewriter.TestRewriter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/izeigerman/github/sqlglot/tests/test_rewriter.py", line 11, in test_ctas
    Rewriter(expression).ctas('x').expression.sql(),
  File "/Users/izeigerman/github/sqlglot/sqlglot/rewriter.py", line 13, in wrapper
    expression = self.func(rewriter, *args, **kwargs)
  File "/Users/izeigerman/github/sqlglot/sqlglot/rewriter.py", line 37, in ctas
    file_format = exp.FileFormat(this=file_format),
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 14, in __init__
    self.validate()
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 81, in validate
    raise ValueError(f"Required keyword: {k} missing for {self.token_type}")
ValueError: Required keyword: this missing for TokenType.FORMAT

======================================================================
ERROR: test_identity (tests.test_transpile.TestTranspile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/izeigerman/github/sqlglot/tests/test_transpile.py", line 181, in test_identity
    self.assertEqual(transpile(sql)[0], sql.strip())
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 24, in transpile
    for expression in parse(code, read, error_level=error_level)
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 13, in parse
    return dialect.parse(code, **opts)
  File "/Users/izeigerman/github/sqlglot/sqlglot/dialects.py", line 17, in parse
    return self.parser(**opts).parse(self.tokenizer().tokenize(code), code)
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 193, in parse
    expressions.append(self._parse_statement())
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 261, in _parse_statement
    cte = self._parse_cte()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 397, in _parse_cte
    return self._parse_select()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 429, in _parse_select
    'from': self._parse_from(),
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 456, in _parse_from
    return exp.From(expressions=self._parse_csv(self._parse_table))
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 941, in _parse_csv
    items = [parse()]
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 530, in _parse_table
    this = exp.Alias(this=this, alias=None)
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 14, in __init__
    self.validate()
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 81, in validate
    raise ValueError(f"Required keyword: {k} missing for {self.token_type}")
ValueError: Required keyword: alias missing for TokenType.ALIAS

======================================================================
ERROR: test_partial (tests.test_transpile.TestTranspile)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/izeigerman/github/sqlglot/tests/test_transpile.py", line 186, in test_partial
    self.assertEqual(transpile(sql, error_level=ErrorLevel.IGNORE)[0], sql.strip())
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 24, in transpile
    for expression in parse(code, read, error_level=error_level)
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 13, in parse
    return dialect.parse(code, **opts)
  File "/Users/izeigerman/github/sqlglot/sqlglot/dialects.py", line 17, in parse
    return self.parser(**opts).parse(self.tokenizer().tokenize(code), code)
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 193, in parse
    expressions.append(self._parse_statement())
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 261, in _parse_statement
    cte = self._parse_cte()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 397, in _parse_cte
    return self._parse_select()
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 429, in _parse_select
    'from': self._parse_from(),
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 456, in _parse_from
    return exp.From(expressions=self._parse_csv(self._parse_table))
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 941, in _parse_csv
    items = [parse()]
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 520, in _parse_table
    expression = exp.Table(this=table, db=db)
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 14, in __init__
    self.validate()
  File "/Users/izeigerman/github/sqlglot/sqlglot/expressions.py", line 81, in validate
    raise ValueError(f"Required keyword: {k} missing for {self.token_type}")
ValueError: Required keyword: this missing for TokenType.TABLE

----------------------------------------------------------------------
Ran 31 tests in 0.213s

FAILED (errors=4)
```

Comment (izeigerman) [reply sqlglot/dialects.py:119]: Perhaps I should do this validation for functions only

Comment (tobymao) [reply sqlglot/dialects.py:119]: can we fix those expressions to not have those args required?

File: sqlglot/dialects.py (hunk 3 above)
Comment (tobymao) [reply sqlglot/dialects.py:229]: makes sense

File: sqlglot/dialects.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/dialects.py:119]: I can try that! Just wasn't sure whether this will break some other assumption. It's hard to assess the impact.

File: sqlglot/dialects.py (hunk 3 above)
Comment (izeigerman) [reply sqlglot/dialects.py:229]: So it seems like we can't really push the validation into this layer. For some dialects we actually want to ignore some arguments while constructing the tree. If we unconditionally fail in the parser we won't be able to support use cases like:
```
DATE_ADD('day', 1, x)
```
for Presto dialect for which we actually ignore the first `'day'` argument. 
@tobymao seems like the validation should still take place in the lambda and subsequently in `from_arg_list`. What do you think?

File: sqlglot/dialects.py (hunk 1 above)
Comment (izeigerman) [reply sqlglot/dialects.py:119]: @tobymao I've refactored the error handling a bit and made it more consistent. Now the following values are treated as missing arguments:
```
None
[]
[None]  # Captured a couple of those as well
```
Please check out the corresponding [commit](https://github.com/tobymao/sqlglot/pull/25/commits/127660b6c94946f20aa6261e05ec13e78474e1a9). Please note that with this change the parsing of badly formed expressions no longer succeeds like it did before. Instead the exception is thrown or no expression is returned if the `ErrorLevel` is set to `IGNORE`.

File: sqlglot/dialects.py (hunk 3 above)
Comment (izeigerman) [reply sqlglot/dialects.py]: Fixed as per our conversation.

File: sqlglot/expressions.py [hunk 7]
```diff
@@ -171,7 +170,7 @@ class CharacterSet(Expression):
 
 class CTE(Expression):
     token_type = TokenType.WITH
-    arg_types = {'this': True, 'expressions': True, 'recursive': False}
+    arg_types = {'this': False, 'expressions': True, 'recursive': False}
```
Comment (izeigerman) [inline sqlglot/expressions.py:173]: @tobymao I'm a little concerned about this change. Had to do this because of:
```python
  File "/Users/izeigerman/github/sqlglot/tests/test_dialects.py", line 517, in test_spark
    transpile(
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 24, in transpile
    for expression in parse(code, read, error_level=error_level)
  File "/Users/izeigerman/github/sqlglot/sqlglot/__init__.py", line 13, in parse
    return dialect.parse(code, **opts)
  File "/Users/izeigerman/github/sqlglot/sqlglot/dialects.py", line 17, in parse
    return self.parser(**opts).parse(self.tokenizer().tokenize(code), code)
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 198, in parse
    self.raise_error(str(e))
  File "/Users/izeigerman/github/sqlglot/sqlglot/parser.py", line 225, in raise_error
    raise self.error
sqlglot.errors.ParseError: Required keyword: 'this' missing for TokenType.WITH. Line 0, Col: 34.
WITH RECURSIVE T(N) AS (VALUES (1))
```

Comment (tobymao) [reply sqlglot/expressions.py:173]: ah a recursive cte... i guess a recursive cte doesn't have the name

Comment (tobymao) [reply sqlglot/expressions.py:173]: shouldn't the this be T(N) in this example?

File: sqlglot/expressions.py [hunk 8]
```diff
@@ -484,10 +483,33 @@ class Interval(Expression):
 # Functions
 class Func(Expression):
     token_type = TokenType.FUNC
+    is_var_len_args = False
+
+    @classmethod
+    def from_arg_list(cls, args):
+        args_num = len(args)
+
+        all_arg_keys = list(cls.arg_types.keys())
```
Comment (tobymao) [inline sqlglot/expressions.py:492]: don't need to call keys, list(cls.arg_types) does the same thing

File: sqlglot/generator.py [hunk 9]
```diff
@@ -540,8 +540,11 @@ def binary(self, expression, op, newline=False):
 
     def expressions(self, expression, flat=False, pad=0):
         # pylint: disable=cell-var-from-loop
+        expressions = expression.args['expressions']
```
Comment (tobymao) [inline sqlglot/generator.py:543]: expressions = expression.get('expressions') or []

File: sqlglot/tokens.py [hunk 10]
```diff
@@ -166,7 +166,7 @@ def number(cls, number):
 
     @classmethod
     def string(cls, string):
-        return cls(TokenType.STRING, f"'{string}'")
+        return cls(TokenType.STRING, f"{string}")
```
Comment (tobymao) [inline sqlglot/tokens.py:169]: this should just be string no need to interpolate it if you're not adding quotes

Comment (izeigerman) [reply sqlglot/tokens.py:169]: ah, good catch. I fixed quotes but forgot the interpolation

File: tests/test_transpile.py [hunk 11]
```diff
@@ -183,7 +183,7 @@ def test_identity(self):
     def test_partial(self):
         with open(os.path.join(self.fixtures_dir, 'partial.sql'), encoding='utf-8') as f:
             for sql in f:
-                self.assertEqual(transpile(sql, error_level=ErrorLevel.IGNORE)[0], sql.strip())
+                self.assertEqual(len(transpile(sql, error_level=ErrorLevel.IGNORE)), 0)
```
Comment (tobymao) [inline tests/test_transpile.py:186]: why was this changed?

Comment (izeigerman) [reply tests/test_transpile.py:186]: Please see my [comment](https://github.com/tobymao/sqlglot/pull/25#discussion_r714214201). Basically since we now treat all `None` arguments as missing parsing of badly formed expressions now fails.

Comment (tobymao) [reply tests/test_transpile.py:186]: ahh... that's a bit of a bummer that that's no longer support, since someone requested this feature

is it possible to allow this to happen (through some flag) -- the use case was so that we could parse partial expressions?

File: sqlglot/expressions.py (hunk 7 above)
Comment (izeigerman) [reply sqlglot/expressions.py:173]: Yeah, I kind of thought it should be `T` or something. Here's the parsed tree:
```
[(CTE this: , expressions:
  (ALIAS this:
    (VALUES expressions:
      (TUPLE expressions: 1)), alias:
    (ANONYMOUS this: T, expressions:
      (COLUMN this: N, db: , table: , fields: ))), recursive: RECURSIVE)]
```
Does this look correct to you?

Comment (izeigerman) [reply sqlglot/expressions.py:173]: Ah, seems like it actually expects `SELECT` and not `VALUES`.

Comment (tobymao) [reply sqlglot/expressions.py:173]: ah, this may be a case of a bad test

`
WITH RECURSIVE t(n) AS (
    VALUES (1)
  UNION ALL
    SELECT n+1 FROM t WHERE n < 100
)
SELECT sum(n) FROM t;`

is valid, my test was probably a partial

Comment (tobymao) [reply sqlglot/expressions.py:173]: makes sense now!

**Review by tobymao — APPROVED** [review]
